### PR TITLE
Data Model Schema Viewer and Mapper

### PR DESCRIPTION
**Pull request by bebinu**
* data model viewer schema definition
* data model schema loader
* data model api handler
* data model mapping directive

### CODE REVIEW

File: wrangler-service/src/main/java/io/cdap/wrangler/service/schema/DataModelHandler.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import javax.ws.rs.PathParam;
+
+/**
+ * The {@link DataModelHandler} class provides data model schema listings APIs.
+ */
+public class DataModelHandler extends AbstractWranglerHandler {
+
+
+  @POST
+  @Path("v2/contexts/{context}/schemas")
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/io/cdap/wrangler/service/schema/DataModelHandler.java]: Other handlers are not using v2 version in the apis. I think it would be good to be consistent with them

File: wrangler-service/src/main/java/io/cdap/wrangler/service/schema/DataModelHandler.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  @TransactionPolicy(value = TransactionControl.EXPLICIT)
+  public void upload(HttpServiceRequest request, HttpServiceResponder responder,
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/io/cdap/wrangler/service/schema/DataModelHandler.java:66]: please add java docs for these methods

File: wrangler-core/src/main/java/io/cdap/wrangler/datamodel/DataModelGlossary.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+package io.cdap.wrangler.datamodel;
+
+import io.cdap.wrangler.utils.AvroSchemaGlossary;
+import io.cdap.wrangler.utils.AvroSchemaLoader;
+import io.cdap.wrangler.utils.PackagedSchemaLoader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Singelton class for accessing an {@link AvroSchemaGlossary}
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/datamodel/DataModelGlossary.java]: typo: singleton

File: wrangler-core/src/main/java/io/cdap/wrangler/datamodel/DataModelGlossary.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ */
+public class DataModelGlossary {
+
+  private static AvroSchemaGlossary glossary;
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/datamodel/DataModelGlossary.java:25]: it should be private static

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/PackagedSchemaLoader.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.file.Paths;
+import java.util.Map;
+
+/**
+ * The {@link PackagedSchemaLoader} object loads an AVRO data model schema bundled as a resource within a package.
+ */
+public class PackagedSchemaLoader implements AvroSchemaLoader {
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/utils/PackagedSchemaLoader.java]: is this only used in tests?

File: wrangler-core/src/main/java/io/cdap/wrangler/datamodel/DataModelGlossary.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.cdap.wrangler.datamodel;
+
+import io.cdap.wrangler.utils.AvroSchemaGlossary;
+import io.cdap.wrangler.utils.AvroSchemaLoader;
+import io.cdap.wrangler.utils.PackagedSchemaLoader;
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/datamodel/DataModelGlossary.java]: nit: unused imports

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/AvroSchemaGlossary.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * The {@link AvroSchemaGlossary} class is used for management of AVRO data model schema definitions.
+ */
+public class AvroSchemaGlossary {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AvroSchemaGlossary.class);
+  private static final String REVISION_PROPERTY = "_revision";
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/utils/AvroSchemaGlossary.java:35]: does it mean schema glossary always should have a revision number?

**Review by CuriousVini — COMMENTED** [review]
Is there any specific reason for not reusing `SchemaRegistryHandler`?

File: wrangler-service/src/main/java/io/cdap/wrangler/service/schema/DataModelHandler.java (hunk 1 above)
Comment (bebinu) [reply wrangler-service/src/main/java/io/cdap/wrangler/service/schema/DataModelHandler.java]: Done.

File: wrangler-service/src/main/java/io/cdap/wrangler/service/schema/DataModelHandler.java (hunk 2 above)
Comment (bebinu) [reply wrangler-service/src/main/java/io/cdap/wrangler/service/schema/DataModelHandler.java:66]: Done.

File: wrangler-core/src/main/java/io/cdap/wrangler/datamodel/DataModelGlossary.java (hunk 3 above)
Comment (bebinu) [reply wrangler-core/src/main/java/io/cdap/wrangler/datamodel/DataModelGlossary.java]: Fixed.

File: wrangler-core/src/main/java/io/cdap/wrangler/datamodel/DataModelGlossary.java (hunk 4 above)
Comment (bebinu) [reply wrangler-core/src/main/java/io/cdap/wrangler/datamodel/DataModelGlossary.java:25]: It is already private static.

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/PackagedSchemaLoader.java (hunk 5 above)
Comment (bebinu) [reply wrangler-core/src/main/java/io/cdap/wrangler/utils/PackagedSchemaLoader.java]: Removed and a stub for testing purposes has been added.

File: wrangler-core/src/main/java/io/cdap/wrangler/datamodel/DataModelGlossary.java (hunk 6 above)
Comment (bebinu) [reply wrangler-core/src/main/java/io/cdap/wrangler/datamodel/DataModelGlossary.java]: Fixed.

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/AvroSchemaGlossary.java (hunk 7 above)
Comment (bebinu) [reply wrangler-core/src/main/java/io/cdap/wrangler/utils/AvroSchemaGlossary.java:35]: The data model schema must always have a _revision meta data property. The schema glossary uses the _revision meta data property and the fullname to uniquely identify each schema.

**Comment by bebinu** [comment]
Thanks for the review!

> Is there any specific reason for not reusing `SchemaRegistryHandler`?
The semantics and use cases for the SchemaRegistryHandler is a bit different than the DataModelHandler.

- the schema within the SchemaRegistryHandler are versioned with an internal auto incremented version that is returned to the client during upload. While the schema within the DataModelHandler maintains a revision derived from the schema.
- the SchemaRegistryHandler expects the client to generate ids and to associate schemas with the ids where as the DataModelHandler uses the full name and revision of a schema as a unique identifier.
- the use case for the SchemaRegistryHandler targets the CRUD operations for management of a client mutating schema. The use case for the DataModelHandler involves the client performing listings of all available data model schemas for selection, viewing, but no editing.

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+ * the License.
+ */
+package io.cdap.wrangler.utils;
+
+import io.cdap.cdap.api.common.Bytes;
+import io.cdap.wrangler.api.DirectiveExecutionException;
+import io.cdap.wrangler.api.Row;
+
+/**
+ * Utility class that converts a {@link Row} column into another column.
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java:23]: we follow coding pattern where a util class should have a private constructor. Directive name can be passed as a parameter to the methods

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+ */
+public final class ColumnConverter {
+  private final String directiveName;
+
+  public ColumnConverter(String directiveName) {
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java]: usually for util classes, we make constructor private. directive name can be an argument to methods.

File: wrangler-core/src/test/java/io/cdap/wrangler/datamodel/HTTPSchemLoaderTest.java [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+import java.util.List;
+import java.util.Set;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+/**
+ * Tests {@link HTTPSchemaLoader}
+ */
+public class HTTPSchemLoaderTest {
```
Comment (CuriousVini) [inline wrangler-core/src/test/java/io/cdap/wrangler/datamodel/HTTPSchemLoaderTest.java]: typo: HTTPSchemaLoaderTest

File: wrangler-core/src/main/java/io/cdap/wrangler/datamodel/HTTPSchemaLoader.java [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * The {@link HTTPSchemaLoader} object loads an AVRO data model schema located in a GCS bucket.
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/datamodel/HTTPSchemaLoader.java]: this will load schema from any url correct?

File: wrangler-core/src/test/java/io/cdap/directives/validation/ValidateStandardTest.java [hunk 12]
```diff
@@ -156,12 +157,6 @@ public void verifyManifest() throws Exception {
           declared.toString(), actual.toString()),
         declared,
         actual);
-
-      assertTrue(
```
Comment (CuriousVini) [inline wrangler-core/src/test/java/io/cdap/directives/validation/ValidateStandardTest.java]: why is this change needed?

File: wrangler-core/src/test/java/io/cdap/directives/validation/ValidateStandardTest.java [hunk 13]
```diff
@@ -18,9 +18,10 @@
 
 import com.google.gson.Gson;
 import com.google.gson.JsonObject;
-import io.cdap.directives.validation.Manifest.Standard;
 import io.cdap.wrangler.TestingRig;
 import io.cdap.wrangler.api.Row;
+import io.cdap.wrangler.utils.Manifest;
```
Comment (CuriousVini) [inline wrangler-core/src/test/java/io/cdap/directives/validation/ValidateStandardTest.java:23]: nit: unused imports?

File: wrangler-service/src/main/java/io/cdap/wrangler/service/directive/DirectivesHandler.java [hunk 14]
```diff
@@ -816,6 +823,173 @@ public void schema(HttpServiceRequest request, HttpServiceResponder responder,
...
 
+  /**
+   * Adds a data model to workspace. There is a one-to-one mapping between workspaces and models.
+   *
+   * @param request Handler for incoming request.
+   * @param responder Responder for data going out.
+   * @param id Workspace to associate a data model.
+   */
+  @POST
+  @Path("contexts/{context}/workspaces/{id}/datamodels")
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/io/cdap/wrangler/service/directive/DirectivesHandler.java:834]: why does this need to be in directives handler? can this be moved to datamodel handler?

**Review by CuriousVini — COMMENTED** [review]
a few questions and minor comments.

File: wrangler-service/src/main/java/io/cdap/wrangler/service/directive/DirectivesHandler.java (hunk 14 above)
Comment (bebinu) [reply wrangler-service/src/main/java/io/cdap/wrangler/service/directive/DirectivesHandler.java:834]: This end point is updating properties of the workspace that will be used to maintain workspace related state client side. Also, following a similar pattern to the schema route within directives handler.

File: wrangler-core/src/test/java/io/cdap/wrangler/datamodel/HTTPSchemLoaderTest.java (hunk 10 above)
Comment (bebinu) [reply wrangler-core/src/test/java/io/cdap/wrangler/datamodel/HTTPSchemLoaderTest.java]: Fixed, thanks.

File: wrangler-core/src/test/java/io/cdap/directives/validation/ValidateStandardTest.java (hunk 12 above)
Comment (bebinu) [reply wrangler-core/src/test/java/io/cdap/directives/validation/ValidateStandardTest.java]: Reverted.

File: wrangler-core/src/test/java/io/cdap/directives/validation/ValidateStandardTest.java (hunk 13 above)
Comment (bebinu) [reply wrangler-core/src/test/java/io/cdap/directives/validation/ValidateStandardTest.java:23]: Used on line 133.

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java (hunk 8 above)
Comment (bebinu) [reply wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java:23]: Done.

File: wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java (hunk 9 above)
Comment (bebinu) [reply wrangler-core/src/main/java/io/cdap/wrangler/utils/ColumnConverter.java]: Done.

File: wrangler-core/src/main/java/io/cdap/wrangler/datamodel/HTTPSchemaLoader.java (hunk 11 above)
Comment (bebinu) [reply wrangler-core/src/main/java/io/cdap/wrangler/datamodel/HTTPSchemaLoader.java]: Updated, done.

**Review by CuriousVini — APPROVED** [review]
lgtm

**Comment by CuriousVini** [comment]
please squash all the commits

**Comment by bebinu** [comment]
> please squash all the commits

Done.

**Comment by bebinu** [comment]
> lgtm

Thanks for the review!

**Comment by CuriousVini** [comment]
Thanks for the contribution @bebinu !!
